File: python/eventgraph/grammar.py
"""Social Grammar — high-level vertex operations on the event graph.

Ports the vertex operations from the Go reference implementation.
Each operation gets the current chain head, creates a properly hash-chained
event, appends it to the store, and returns the event.
"""

from __future__ import annotations

from typing import Sequence

from .event import Signer, create_event
from .store import Store
from .types import ActorID, ConversationID, EventID, EventType, Hash


class Grammar:
    """Wraps a Store and provides high-level social grammar operations.

    Each method creates a properly hash-chained event with the right type,
    content, and causal links, then appends it to the store.
    """

    def __init__(self, store: Store) -> None:
        self._store = store

    def _prev_hash(self) -> Hash:
        """Get the prev_hash for the next event in the chain."""
        head = self._store.head()
        return head.unwrap().hash if head.is_some() else Hash.zero()

    def emit(
        self,
        source: ActorID,
        body: str,
        conversation_id: ConversationID,
        causes: list[EventID],
        signer: Signer,
    ) -> "Event":
        """Emit a new statement into the graph.

        Causes must be non-empty (enforced by create_event / NonEmpty).
        """
        if not causes:
            raise ValueError("emit requires at least one cause")
        prev_hash = self._prev_hash()
        return self._store.append(create_event(
            event_type=EventType("grammar.emitted"),
            source=source,
            content={"Body": body},
            causes=causes,
            conversation_id=conversation_id,
            prev_hash=prev_hash,
            signer=signer,
        ))

    def respond(
        self,
        source: ActorID,
        body: str,
        parent: EventID,
        conversation_id: ConversationID,
        signer: Signer,
    ) -> "Event":
        """Respond to an existing event."""
        prev_hash = self._prev_hash()
        return self._store.append(create_event(
            event_type=EventType("grammar.responded"),
            source=source,
            content={"Body": body, "Parent": parent.value},
            causes=[parent],
            conversation_id=conversation_id,
            prev_hash=prev_hash,
            signer=signer,
        ))

    def derive(
        self,
        source: ActorID,
        body: str,
        source_event: EventID,
        conversation_id: ConversationID,
        signer: Signer,
    ) -> "Event":
        """Derive a new insight from an existing event."""
        prev_hash = self._prev_hash()
        return self._store.append(create_event(
            event_type=EventType("grammar.derived"),
            source=source,
            content={"Body": body, "Source": source_event.value},
            causes=[source_event],
            conversation_id=conversation_id,
            prev_hash=prev_hash,
            signer=signer,
        ))

    def extend(
        self,
        source: ActorID,
        body: str,
        previous: EventID,
        conversation_id: ConversationID,
        signer: Signer,
    ) -> "Event":
        """Extend an existing event with additional content."""
        prev_hash = self._prev_hash()
        return self._store.append(create_event(
            event_type=EventType("grammar.extended"),
            source=source,
            content={"Body": body, "Previous": previous.value},
            causes=[previous],
            conversation_id=conversation_id,
            prev_hash=prev_hash,
            signer=signer,
        ))

    def retract(
        self,
        source: ActorID,
        target: EventID,
        reason: str,
        conversation_id: ConversationID,
        signer: Signer,
    ) -> "Event":
        """Retract a previously emitted event.

        The source must be the original author of the target event.
        Raises ValueError if the source did not author the target.
        """
        target_event = self._store.get(target)
        if target_event.source.value != source.value:
            raise ValueError(
                f"actor {source.value} cannot retract event authored by {target_event.source.value}"
            )
        prev_hash = self._prev_hash()
        return self._store.append(create_event(
            event_type=EventType("grammar.retracted"),
            source=source,
            content={"Target": target.value, "Reason": reason},
            causes=[target],
            conversation_id=conversation_id,
            prev_hash=prev_hash,
            signer=signer,
        ))

    def annotate(
        self,
        source: ActorID,
        target: EventID,
        key: str,
        value: str,
        conversation_id: ConversationID,
        signer: Signer,
    ) -> "Event":
        """Annotate an existing event with a key-value pair."""
        prev_hash = self._prev_hash()
        return self._store.append(create_event(
            event_type=EventType("grammar.annotated"),
            source=source,
            content={"Target": target.value, "Key": key, "Value": value},
            causes=[target],
            conversation_id=conversation_id,
            prev_hash=prev_hash,
            signer=signer,
        ))

    def merge(
        self,
        source: ActorID,
        body: str,
        sources: Sequence[EventID],
        conversation_id: ConversationID,
        signer: Signer,
    ) -> "Event":
        """Merge multiple events into a synthesis.

        Requires at least 2 source events.
        """
        if len(sources) < 2:
            raise ValueError("merge requires at least 2 source events")
        prev_hash = self._prev_hash()
        return self._store.append(create_event(
            event_type=EventType("grammar.merged"),
            source=source,
            content={"Body": body, "Sources": [s.value for s in sources]},
            causes=list(sources),
            conversation_id=conversation_id,
            prev_hash=prev_hash,
            signer=signer,
        ))

File: dotnet/src/EventGraph/Grammar.cs
namespace EventGraph;

/// <summary>
/// Social grammar providing high-level vertex operations that create properly
/// hash-chained events on an IStore. Ports the vertex operations from the Go
/// reference implementation.
/// </summary>
public sealed class Grammar
{
    private readonly IStore _store;

    public Grammar(IStore store) => _store = store;

    /// <summary>Creates independent content. Requires at least one cause.</summary>
    public Event Emit(ActorId source, string body, ConversationId conversationId, List<EventId> causes, ISigner signer)
    {
        if (causes.Count == 0)
            throw new ArgumentException("Emit requires at least one cause");

        var head = _store.Head();
        var prevHash = head.IsSome ? head.Unwrap().Hash : Hash.Zero();
        return _store.Append(EventFactory.CreateEvent(
            new EventType("grammar.emitted"), source,
            new Dictionary<string, object?> { ["Body"] = body },
            causes, conversationId, prevHash, signer));
    }

    /// <summary>Creates causally dependent, subordinate content.</summary>
    public Event Respond(ActorId source, string body, EventId parent, ConversationId conversationId, ISigner signer)
    {
        var head = _store.Head();
        var prevHash = head.IsSome ? head.Unwrap().Hash : Hash.Zero();
        return _store.Append(EventFactory.CreateEvent(
            new EventType("grammar.responded"), source,
            new Dictionary<string, object?> { ["Body"] = body, ["Parent"] = parent.Value },
            new List<EventId> { parent }, conversationId, prevHash, signer));
    }

    /// <summary>Creates causally dependent but independent content.</summary>
    public Event Derive(ActorId source, string body, EventId sourceEvent, ConversationId conversationId, ISigner signer)
    {
        var head = _store.Head();
        var prevHash = head.IsSome ? head.Unwrap().Hash : Hash.Zero();
        return _store.Append(EventFactory.CreateEvent(
            new EventType("grammar.derived"), source,
            new Dictionary<string, object?> { ["Body"] = body, ["Source"] = sourceEvent.Value },
            new List<EventId> { sourceEvent }, conversationId, prevHash, signer));
    }

    /// <summary>Creates sequential content from the same author.</summary>
    public Event Extend(ActorId source, string body, EventId previous, ConversationId conversationId, ISigner signer)
    {
        var head = _store.Head();
        var prevHash = head.IsSome ? head.Unwrap().Hash : Hash.Zero();
        return _store.Append(EventFactory.CreateEvent(
            new EventType("grammar.extended"), source,
            new Dictionary<string, object?> { ["Body"] = body, ["Previous"] = previous.Value },
            new List<EventId> { previous }, conversationId, prevHash, signer));
    }

    /// <summary>Tombstones own content. Only the original author can retract.</summary>
    public Event Retract(ActorId source, EventId target, string reason, ConversationId conversationId, ISigner signer)
    {
        var targetEvent = _store.Get(target);
        if (targetEvent.Source != source)
            throw new InvalidOperationException(
                $"Actor {source.Value} cannot retract event {target.Value} authored by {targetEvent.Source.Value}");

        var head = _store.Head();
        var prevHash = head.IsSome ? head.Unwrap().Hash : Hash.Zero();
        return _store.Append(EventFactory.CreateEvent(
            new EventType("grammar.retracted"), source,
            new Dictionary<string, object?> { ["Target"] = target.Value, ["Reason"] = reason },
            new List<EventId> { target }, conversationId, prevHash, signer));
    }

    /// <summary>Attaches metadata to existing content.</summary>
    public Event Annotate(ActorId source, EventId target, string key, string value, ConversationId conversationId, ISigner signer)
    {
        var head = _store.Head();
        var prevHash = head.IsSome ? head.Unwrap().Hash : Hash.Zero();
        return _store.Append(EventFactory.CreateEvent(
            new EventType("grammar.annotated"), source,
            new Dictionary<string, object?> { ["Target"] = target.Value, ["Key"] = key, ["Value"] = value },
            new List<EventId> { target }, conversationId, prevHash, signer));
    }

    /// <summary>Joins two or more independent subtrees. Requires at least two sources.</summary>
    public Event Merge(ActorId source, string body, List<EventId> sources, ConversationId conversationId, ISigner signer)
    {
        if (sources.Count < 2)
            throw new ArgumentException("Merge requires at least two sources");

        var head = _store.Head();
        var prevHash = head.IsSome ? head.Unwrap().Hash : Hash.Zero();
        return _store.Append(EventFactory.CreateEvent(
            new EventType("grammar.merged"), source,
            new Dictionary<string, object?> { ["Body"] = body, ["Sources"] = sources.Select(s => (object?)s.Value).ToList() },
            sources, conversationId, prevHash, signer));
    }
}
